File: C++/happy-number.cpp
/*
 * HAPPY NUMBER ALGORITHM IMPLEMENTATION
 * 
 * This file demonstrates the Happy Number algorithm with multiple approaches:
 * - Mathematical problem solving
 * - Cycle detection algorithms
 * - Hash set usage for tracking
 * - Floyd's cycle detection algorithm
 * - Input validation and user interaction
 * 
 * Learning Objectives:
 * - Understand the Happy Number mathematical concept
 * - Learn cycle detection algorithms
 * - Master hash set usage for tracking visited numbers
 * - Apply Floyd's cycle detection technique
 * - Develop problem-solving skills with mathematical algorithms
 * 
 * Prerequisites:
 * - Basic C++ knowledge
 * - Understanding of loops and conditionals
 * - Familiarity with data structures (unordered_set)
 * - Basic understanding of mathematical algorithms
 * 
 * Key Topics Covered:
 * - Happy Number algorithm
 * - Cycle detection techniques
 * - Hash set operations
 * - Floyd's cycle detection
 * - Mathematical problem solving
 */

/*
 * PROBLEM DESCRIPTION
 * 
 * A happy number is a number defined by the following process:
 * 1. Starting with any positive integer, replace the number by the sum of the squares of its digits
 * 2. Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle
 * 3. Those numbers for which this process ends in 1 are happy
 * 
 * Examples:
 * - 19 is happy: 1² + 9² = 82 → 8² + 2² = 68 → 6² + 8² = 100 → 1² + 0² + 0² = 1
 * - 2 is not happy: enters a cycle that doesn't include 1
 */

#include <iostream>      // For input/output operations (cout, cin)
#include <unordered_set> // For hash set data structure

using namespace std; // Using standard namespace

/*
 * HELPER FUNCTION: SUM OF SQUARES OF DIGITS
 * 
 * This function calculates the sum of squares of all digits in a number.
 * It extracts each digit using modulo and division operations.
 * 
 * @param n The number to process
 * @return The sum of squares of all digits
 */
int sumOfSquares(int n) {
    int sum = 0;  // Initialize sum to 0
    
    // Process each digit of the number
    while (n > 0) {
        int digit = n % 10;        // Extract the last digit
        sum += digit * digit;      // Add square of digit to sum
        n /= 10;                   // Remove the last digit
    }
    
    return sum;  // Return the calculated sum
}

/*
 * HAPPY NUMBER DETECTION USING HASH SET
 * 
 * This function determines if a number is happy using a hash set to detect cycles.
 * It keeps track of all numbers seen during the process to detect infinite loops.
 * 
 * @param n The number to check
 * @return true if the number is happy, false otherwise
 */
bool isHappy(int n) {
    unordered_set<int> seen;  // Hash set to track seen numbers
    
    // Continue until we reach 1 or detect a cycle
    while (n != 1 && !seen.count(n)) {
        seen.insert(n);           // Mark current number as seen
        n = sumOfSquares(n);      // Calculate next number in sequence
    }
    
    return n == 1;  // Return true if we reached 1
}

/*
 * HAPPY NUMBER DETECTION USING FLOYD'S CYCLE DETECTION
 * 
 * This function uses Floyd's cycle detection algorithm (tortoise and hare) to detect cycles.
 * It uses two pointers moving at different speeds to detect cycles without extra space.
 * 
 * @param n The number to check
 * @return true if the number is happy, false otherwise
 */
bool isHappyFloyd(int n) {
    int slow = n;                    // Slow pointer (tortoise)
    int fast = sumOfSquares(n);      // Fast pointer (hare)
    
    // Move pointers until they meet or reach 1
    while (fast != 1 && fast != slow) {
        slow = sumOfSquares(slow);                    // Move slow pointer 1 step
        fast = sumOfSquares(sumOfSquares(fast));      // Move fast pointer 2 steps
    }
    
    return fast == 1;  // Return true if fast pointer reached 1
}

/*
 * MAIN FUNCTION - HAPPY NUMBER CHECKER
 * 
 * This function provides an interactive interface for checking if a number is happy.
 * It demonstrates both approaches to solving the happy number problem.
 */
int main() {
    int num;  // Variable to store user input
    
    // Get number from user
    cout << "Enter a number: ";
    cin >> num;
    
    // Check if the number is happy using hash set approach
    if (isHappy(num)) {
        cout << num << " is a happy number." << endl;
    } else {
        cout << num << " is not a happy number." << endl;
    }
    
    // Return 0 to indicate successful program execution
    return 0;
}

/*

#include <iostream>

using namespace std;

int sumOfSquares(int n) {
    int sum = 0;
    while (n > 0) {
        int digit = n % 10;
        sum += digit * digit;
        n /= 10;
    }
    return sum;
}

bool isHappy(int n) {
    int slow = n;
    int fast = sumOfSquares(n);

    while (fast != 1 && slow != fast) {
        slow = sumOfSquares(slow);           // move 1 step
        fast = sumOfSquares(sumOfSquares(fast)); // move 2 steps
    }

    return fast == 1;
}

int main() {
    int num;
    cout << "Enter a number: ";
    cin >> num;

    if (isHappy(num)) {
        cout << num << " is a happy number." << endl;
    } else {
        cout << num << " is not a happy number." << endl;
    }

    return 0;
}

*/

File: Python/01_beginner_level/happy_number.py
'''
Write an algorithm to determine if a number n is happy.

A happy number is a number defined by the following process:

Starting with any positive integer, replace the number by the sum of the squares of its digits.
Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
Those numbers for which this process ends in 1 are happy.
Return true if n is a happy number, and false if not.

Example 1:
Input: n = 19, 62387879
Output: true
Explanation:
1^2 + 9^2 = 82 => 9^2 + 1^2 = 82
8^2 + 2^2 = 68
6^2 + 8^2 = 100
1^2 + 0^2 + 0^2 = 1
Return true because it ends in 1.

62387879 % 10 = 9
62387879 // 10 = 6238787
6238787 % 10 = 7
6238787 // 10 = 623878
623878 % 10 = 8
623878 // 10 = 62387
62387 % 10 = 7
62387 // 10 = 6238
6238 % 10 = 8
6238 // 10 = 623
623 % 10 = 3
623 // 10 = 62
62 % 10 = 2
62 // 10 = 6
6 % 10 = 6
6 // 10 = 0


Example 2:
Input: n = 2
Output: false
Explanation:
2^2 = 4
4^2 = 16
1^2 + 6^2 = 37
3^2 + 7^2 = 58
5^2 + 8^2 = 89
8^2 + 9^2 = 145
1^2 + 4^2 + 5^2 = 42
4^2 + 2^2 = 20
2^2 + 0^2 = 4
Return false because it loops endlessly in a cycle which does not include 1.
'''
from typing import Any


from ast import Set


def getNextNumber(number: int) -> int:
    nextNumber = 0
    while number > 0:
        nextNumber += (number % 10) ** 2
        number //= 10
    return nextNumber

def isHappyNumber(number: int) -> bool:
    seenNumbers = set[Any]() # to store the numbers we have seen
    while number != 1 and number not in seenNumbers:
        seenNumbers.add(number)
        number = getNextNumber(number)
    return number == 1

### Floyd's Cycle Detection Algorithm
def isHappyNumber(number: int) -> bool:
    slow = number
    fast = number
    while fast != 1 and fast != slow:
        slow = getNextNumber(slow)
        fast = getNextNumber(getNextNumber(fast))
    return fast == 1

